src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -187,18 +187,34 @@ VideoDecoder::VideoStreamDecoderOptions::VideoStreamDecoderOptions(
   }
 }
 
-VideoDecoder::BatchDecodedOutput::BatchDecodedOutput(
-    int64_t numFrames,
-    const VideoStreamDecoderOptions& options,
-    const StreamMetadata& metadata)
-    : frames(torch::empty(
-          {numFrames,
-           options.height.value_or(*metadata.height),
-           options.width.value_or(*metadata.width),
-           3},
-          at::TensorOptions(options.device).dtype(torch::kUInt8))),
-      ptsSeconds(torch::empty({numFrames}, {torch::kFloat64})),
-      durationSeconds(torch::empty({numFrames}, {torch::kFloat64})) {}
+torch::Tensor VideoDecoder::allocateEmptyHWCTensorForStream(
+    int streamIndex,
```

Comment:
Please double check this:

To allocate a tensor, we need its height and width. To get those, we need a streamIndex[1]. To get the streamIndex, we need either:
- it to be passed as a parameter
- for the "NoDemux" functions, we need at least a `RawOutput`.


[1] well, not strictly true, there are some cases where it's not needed, but the whole point of this PR is to unify that logic.

Reply:
In the most generic case this isn't true because a single stream can have frames with different heights and widths:

https://github.com/pytorch/torchcodec/issues/312

But ignoring that fact, to get the frame dimensions before decoding you do need the stream_index